### ISSUE TITLE
An ip address of a new client

### ISSUE THREAD

**Issue by B0mbo** [zaphoyd/websocketpp#694]
Hello!
How can I get an ip address of a new client connection?

**Comment by gentzel** [comment]
Untested, but this should work (I do something very similar to get remote port).  In your open handler:
```
const auto con = endpoint.get_con_from_hdl(hdl);
// ...
const auto address = con->get_raw_socket().remote_endpoint().address();
```

**Comment by qwertie** [comment]
For anyone else who is wondering how this code works and the types involved...

- get_con_from_hdl is defined in `websocketpp::endpoint<connection,config>` (websocketpp/endpoint.hpp), the base class of `websocketpp::server<config>`. It returns `connection_ptr` which resolves to a shared ptr to `websocketpp::connection<config>`.
- `websocketpp::connection<config>` doesn't have a `get_raw_socket` method, but it has three base classes, and the first one is `config::transport_type::transport_con_type`. If you're using Asio TLS, `config::transport_type` resolves to `websocketpp::transport::asio::endpoint<transport_config>`, and the base class resolves to to,`websocketpp::transport::asio::connection<websocketpp::config::asio_tls::transport_config>`. But that class doesn't have a `get_raw_socket` method either.
- But, the base class also has a base class, `config::socket_type::socket_con_type`. If `config` is `websocketpp::config::asio_tls::transport_config`, this resolves to `websocketpp::transport::asio::tls_socket::endpoint::socket_con_type` which resolves to `websocketpp::transport::asio::tls_socket::connection`... Finally, this base class has the `get_raw_socket` method.
- `get_raw_socket` returns `socket_type::lowest_layer_type&` which resolves to `boost::asio::ssl::stream<boost::asio::ip::tcp::socket>::lowest_layer_type` which resolves to `boost::asio::basic_socket<tcp>`.

This is far from the most intuitive library I've used - there are three classes that are bases of each other which end up all having the same name, `connection`, but the actual declared names of these base classes are quite different - `transport_con_type` and `socket_con_type`. In hindsight, `con` obviously means "some sort of class named `connection`".

Another way I could express the types involved is to pick apart the call sequence like so:

~~~cpp
// Secure server
std::shared_ptr<websocketpp::server<websocketpp::config::asio_tls>> tls_server;
boost::shared_ptr<websocketpp::connection<websocketpp::config::asio_tls> > con = tls_server->get_con_from_hdl(hdl);
const boost::asio::basic_socket<boost::asio::ip::tcp>& raw_socket = con->get_raw_socket();
const boost::asio::ip::basic_endpoint<boost::asio::ip::tcp>& rep = raw_socket.remote_endpoint();
boost::asio::ip::address address = rep.address();

// Unencrypted server
std::shared_ptr<websocketpp::server<websocketpp::config::asio>> insecure_server;
boost::shared_ptr<websocketpp::connection<websocketpp::config::asio> > con = insecure_server->get_con_from_hdl(hdl);
const boost::asio::basic_stream_socket<boost::asio::ip::tcp>& raw_socket = con->get_raw_socket();
const boost::asio::ip::basic_endpoint<boost::asio::ip::tcp>& rep = raw_socket.remote_endpoint();
boost::asio::ip::address address = rep.address();
~~~

**Comment by zaphoyd** [comment]
Closing this as the question seems fairly thoroughly answered. Thank you @gentzel and @qwertie .

Regarding why the type system is the way it is, the library uses a policy based design pattern. This allow flexibly composing endpoints with low/zero abstraction overhead. The components that ship with the library can be put together to produce an unencrypted endpoint that processes raw bytes from an iostream as well as a TLS encrypted, deflate compressed endpoint that uses Asio to handle all I/O or any combination of those parts (unencrypted asio, deflate compressed iostream, etc). In each case only the dependencies (openssl, zlib, asio, etc) that are actually used will be included. End users can write their own policies as well that slot in with zero overhead just like the bundled policies.

The downside to this method is that there isn't a single neat "connection" class, but rather each policy has "the parts that it contributes to the connection class". The member functions avaliable to a particular user depend what policies they have chosen.

To that end, I've tried to be consistent with that strategy in naming: `websocketpp::connection` is the core connection class and the functionality that other policies contribute to the composed connection class are all called `websocketpp::<policy namespace>::connection`. For example, `websocketpp::transport::asio::connection` contains all the code added to `websocketpp::connection` when you choose the Asio transport policy and `websocketpp::transport::iostream::connection` contains the code added when you choose that policy instead. Code for `websocketpp::transport::asio::connection` lives in `websocketpp/transport/asio/connection.hpp` and so on. You (and your autocompleting IDE) can determine exactly which functions and types are avaliable by enumerating all provided by the policies you have chosen.
